### PR TITLE
Router and Auth0

### PR DESCRIPTION
**Pull request by tym-xqo**

### CODE REVIEW

File: src/router.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      path: '/callback',
+      name: 'callback',
+      component: Callback
+    }
+  ]
+})
+
+// very basic global guard
+router.beforeEach((to, from, next) => {
+  if (to.name == 'callback') { // check if "to"-route is "callback" and allow access
```
Comment (tym-xqo) [inline src/router.js:26]: I believe this is necessary so Auth0 can always reach, even if user is supposed to be blocked.

File: src/views/Callback.vue [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<template>
+  <div class="callback">Callback</div>
+</template>
+
+<script>
+export default {
+  name: 'callback',
+  mounted() {
+    this.$auth.handleAuthentication().then((data) => {
+      this.$router.push({ name: 'home' })
```
Comment (tym-xqo) [inline src/views/Callback.vue:10]: Despite the handleAuthentication() call here, I don't think there's any real test that the user successfully authenticates. The router.push() happens regardless 😕

File: src/auth.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import auth0 from 'auth0-js'
+import Vue from 'vue'
+
+// exchange the object with your own from the setup step above.
+let webAuth = new auth0.WebAuth({
```
Comment (luzmcosta) [inline src/auth.js]: Abstract configuration to a `config` object?

File: src/views/Callback.vue (hunk 2 above)
Comment (tym-xqo) [reply src/views/Callback.vue:10]: Resolved with router-view tag in App.vue

File: src/auth.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  // TODO: get redirectUri from environment variable
+  // redirectUri: 'http://localhost:8080/callback',
+  redirectUri: 'https://nerium-static.ty-m.pw/callback',
+  // we will use the api/v2/ to access the user information as payload
+  audience: 'https://tym-oao.auth0.com/api/v2/',
+  responseType: 'token id_token',
+  scope: 'openid profile'
+})
+
+let auth = new Vue({
```
Comment (luzmcosta) [inline src/auth.js]: Is this reset later on, or can it be `const` instead of `let`?

File: src/auth.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  computed: {
+    token: {
+      get: function() {
+        return localStorage.getItem('id_token')
```
Comment (luzmcosta) [inline src/auth.js]: I've always been a fan of local storage. 😄 👍 Very nice use of it in this file. ✨

File: src/auth.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    login() {
+      webAuth.authorize()
+    },
+    logout() {
+      return new Promise((resolve, reject) => { 
+        localStorage.removeItem('access_token')
+        localStorage.removeItem('id_token')
+        localStorage.removeItem('expires_at')
+        localStorage.removeItem('user')
+        webAuth.authorize()
```
Comment (luzmcosta) [inline src/auth.js]: Should this call `this.login()` instead of `webAuth.authorize()`?

File: src/auth.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      })
+    },
+    isAuthenticated() {
+      return new Date().getTime() < this.expiresAt
+    },
+    handleAuthentication() {
+      return new Promise((resolve, reject) => {  
+        webAuth.parseHash((err, authResult) => {
+
+          if (authResult && authResult.accessToken && authResult.idToken) {
```
Comment (luzmcosta) [inline src/auth.js]: 🤔 Total suggestion; take it or leave it:

```javascript
const valid = authResult && authResult.accessToken && authResult.idToken;

if (valid) { /* ... */ }
```

File: src/router.js (hunk 1 above)
Comment (luzmcosta) [reply src/router.js:26]: Ohhh. Gotcha.

File: src/views/Callback.vue (hunk 2 above)
Comment (luzmcosta) [reply src/views/Callback.vue:10]: 🤔

**Review by luzmcosta — APPROVED** [review]
Loving this. ✨

File: src/auth.js (hunk 3 above)
Comment (tym-xqo) [reply src/auth.js]: Yeah, good idea.

File: src/auth.js (hunk 7 above)
Comment (tym-xqo) [reply src/auth.js]: Nice, thanks!

File: src/auth.js (hunk 6 above)
Comment (tym-xqo) [reply src/auth.js]: This was all copypasta, but now that you mention it, `webAuth` actually has a built-in `logout()` method. I wonder if we should be using that here instead?

File: src/auth.js (hunk 5 above)
Comment (tym-xqo) [reply src/auth.js]: Yeah, most Auth0 docs and examples suggest using local storage for keeping their credential payloads. Vuex might also be an option here, but I would say only as this iterates up to more complex applications.

File: src/auth.js (hunk 6 above)
Comment (tym-xqo) [reply src/auth.js]: Nope, logout() is different: https://auth0.com/docs/libraries/auth0js/v9#logout

I did go with `this.login()` here.

File: src/auth.js (hunk 7 above)
Comment (tym-xqo) [reply src/auth.js]: Done!

File: src/auth.js (hunk 4 above)
Comment (tym-xqo) [reply src/auth.js]: `const` works

File: src/auth.js (hunk 3 above)
Comment (tym-xqo) [reply src/auth.js]: Kept as TODO for now, but won't be big change.

**Pull request merged**